pkg/apis/provisioning/v1alpha1/provisioner_validation.go

```diff
@@ -12,25 +12,122 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// +kubebuilder:webhook:path=/validate-provisioning-karpenter-sh-v1alpha1-provisioner,mutating=false,sideEffects=None,failurePolicy=fail,groups=provisioning.karpenter.sh,resources=provisioners,verbs=create;update,versions=v1alpha1,name=validation.provisioning.karpenter.sh
+
 package v1alpha1
 
 import (
+	"fmt"
+
+	f "github.com/awslabs/karpenter/pkg/utils/functional"
 	runtime "k8s.io/apimachinery/pkg/runtime"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
-// +kubebuilder:webhook:path=/validate-provisioning-karpenter-sh-v1alpha1-provisioner,mutating=false,sideEffects=None,failurePolicy=fail,groups=provisioning.karpenter.sh,resources=provisioners,verbs=create;update,versions=v1alpha1,name=vprovisioner.kb.io
-
 var _ webhook.Validator = &Provisioner{}
 
-func (r *Provisioner) ValidateCreate() error {
+func (p *Provisioner) ValidateCreate() error {
+	return p.validate()
+}
+
+func (p *Provisioner) ValidateUpdate(old runtime.Object) error {
+	return p.validate()
+}
+
+func (p *Provisioner) ValidateDelete() error {
 	return nil
 }
 
-func (r *Provisioner) ValidateUpdate(old runtime.Object) error {
+func (p *Provisioner) validate() error {
+	return f.AllSucceed(
+		p.validateClusterSpec,
+		p.validateRestrictedLabels,
+		p.validateArchitecture,
+		p.validateOperatingSystem,
+	)
+}
+
+func (p *Provisioner) validateClusterSpec() error {
+	if p.Spec.Cluster == nil {
+		return fmt.Errorf("spec.cluster must be defined")
+	}
+	if len(p.Spec.Cluster.Name) == 0 {
+		return fmt.Errorf("spec.cluster.name cannot be empty")
+	}
+	if len(p.Spec.Cluster.Endpoint) == 0 {
+		return fmt.Errorf("spec.cluster.endpoint cannot be empty")
+	}
+	if len(p.Spec.Cluster.CABundle) == 0 {
+		return fmt.Errorf("spec.cluster.caBundle cannot be empty")
+	}
 	return nil
 }
 
-func (r *Provisioner) ValidateDelete() error {
+func (p *Provisioner) validateRestrictedLabels() error {
+	for label := range p.Spec.Labels {
+		for _, restricted := range RestrictedLabels {
+			if restricted == label {
+				return fmt.Errorf("spec.labels contains restricted label %s", label)
```

Comment:
optional: might be nice to use multierr so they can find out about all the bad labels in one go

Reply:
done